cupy/fft/_callback.pyx

```diff
@@ -0,0 +1,465 @@
+from cupy_backends.cuda.api.runtime cimport _is_hip_environment
+from cupy.core.core cimport ndarray
+from cupy.cuda.device cimport get_compute_capability
+
+import contextlib
+import hashlib
+import importlib
+import os
+import shutil
+import string
+import subprocess
+import sys
+import sysconfig
+import tempfile
+import threading
+
+from cupy import __version__ as _cupy_ver
+from cupy._environment import (get_nvcc_path, get_cuda_path)
+from cupy.cuda.compiler import (_get_bool_env_variable, CompileException)
+# for some reason we can't cimport stuff from cupy.cuda.cufft...
+from cupy.cuda.cufft import (CUFFT_C2C, CUFFT_C2R, CUFFT_R2C,
+                             CUFFT_Z2Z, CUFFT_Z2D, CUFFT_D2Z,
+                             CUFFT_CB_LD_COMPLEX, CUFFT_CB_LD_COMPLEX_DOUBLE,
+                             CUFFT_CB_LD_REAL, CUFFT_CB_LD_REAL_DOUBLE,
+                             CUFFT_CB_ST_COMPLEX, CUFFT_CB_ST_COMPLEX_DOUBLE,
+                             CUFFT_CB_ST_REAL, CUFFT_CB_ST_REAL_DOUBLE,)
+from cupy.cuda.cufft import getVersion as get_cufft_version
+
+
+cdef inline _set_cupy_paths():
+    # Workaround: older Cython cannot use __file__ in global scope
+    global _cupy_root, _cupy_include, _source_dir
+    if _cupy_root is None:
+        _cupy_root = os.path.join(os.path.dirname(__file__), '..')
+        _cupy_include = _cupy_root + '/core/include'
+        _source_dir = _cupy_root + '/cuda/'
+
+
+cdef inline _set_nvcc_path():
+    # get_nvcc_path() could be a long string like "ccache nvcc ..."
+    cdef str nvcc = None
+    global _nvcc
+    if _nvcc == []:
+        nvcc = get_nvcc_path()
+        if nvcc is not None:
+            _nvcc = nvcc.split(' ')
+        else:
+            _nvcc = None
+
+
+# information needed for building an external module
+cdef list _cc = sysconfig.get_config_var('CXX').split(' ')
+cdef str _python_include = sysconfig.get_path('include')
+cdef list _nvcc = []
+cdef str _cuda_path = get_cuda_path()
+cdef str _cuda_include = None  # workaround for Read the Docs...
+if _cuda_path is not None:
+    _cuda_include = _cuda_path + '/include/'
+cdef str _build_ver = str(get_build_version())
+cdef int _cufft_ver = get_cufft_version()
+cdef str _cupy_root = None
+cdef str _cupy_include = None
+cdef str _source_dir = None
+cdef str _ext_suffix = sysconfig.get_config_var('EXT_SUFFIX')
+
+# callback related stuff
+cdef str _callback_dev_code = None
+cdef str _callback_cache_dir = os.environ.get(
+    'CUPY_CACHE_DIR', os.path.expanduser('~/.cupy/callback_cache')) + '/'
+cdef dict _callback_mgr = {}  # keep the Python modules alive
+cdef object _callback_thread_local = threading.local()
+
+
+cdef class _ThreadLocal:
+    cdef _CallbackManager _current_cufft_callback
+
+    def __init__(self):
+        self._current_cufft_callback = None
+
+    @staticmethod
+    cdef _ThreadLocal get():
+        cdef _ThreadLocal tls
+        tls = getattr(_callback_thread_local, 'tls', None)
```

Comment:
This is usually done using try/except AttributeError in other places of the codebase due to speed. (I am not sure if it's noticeable enough) so maybe let's change this to keep consistency?

Reply:
Done in bf7efaa.